### PR TITLE
[Merged by Bors] - fix bug where different beacon values are calculated for the same epoch

### LINKED ISSUE #2781: tortoise beacon should verify that the node is eligible to send messages before sending them
## Description

Tortoise beacon always sends messages even if node is not eligible to send them, or the current state of the node doesn't allow to verify eligibility.

```
1083:{"L":"ERROR","T":"2021-09-14T07:07:10.002+0300","N":"d402d.tBeacon      ","M":"failed to handle proposal message","node_id":"d402d6a59d870552a7d9c7ca6af2d1fada8e189fea82a712aa5495b6019388c0","module":"tBeacon","sessionId":"21eddf44-842e-46b7-9f58-9ea2ce57a531","sender":"J1NuNjpGNRrHU38bq4MPAQw8p28iaR55xGGhiH6JA54A","message":"{\"EpochID\":34,\"NodeID\":{\"Key\":\"d402d6a59d870552a7d9c7ca6af2d1fada8e189fea82a712aa5495b6019388c0\",\"VRFPublicKey\":\"ZPmhxaNNibR4Z08pMYG+AulMqkuL5oiN9zB7mt3ty+Y=\"},\"VRFSignature\":\"E0gkWwRH+JzC8FWjpVQD/k1n/qcWex+8C2wBxGTBxdyZWPgtKdszBPiXDb2P7ZfBU/V1Ar/m3sl3AULR3W+JDA==\"}","errmsg":"get node ATXID for epoch (miner ID d402d): atx for node d402d with publication epoch 33: leveldb: not found","name":"tBeacon"}
1195:{"L":"ERROR","T":"2021-09-14T07:07:40.002+0300","N":"d402d.tBeacon      ","M":"failed to handle first voting message","node_id":"d402d6a59d870552a7d9c7ca6af2d1fada8e189fea82a712aa5495b6019388c0","module":"tBeacon","requestId":"15237178-0aac-411e-8154-0eb8f678de04","gossip_listener_channel":"TBFirstVotingGossip","sessionId":"2c118dcc-a786-4194-b9a4-dd5d55a5c89a","errmsg":"verify first voting message: get node ATXID for epoch (miner ID d402d): atx for node d402d with publication epoch 33: leveldb: not found","name":"tBeacon"}
1326:{"L":"ERROR","T":"2021-09-14T07:08:50.004+0300","N":"d402d.tBeacon      ","M":"failed to handle following voting message","node_id":"d402d6a59d870552a7d9c7ca6af2d1fada8e189fea82a712aa5495b6019388c0","module":"tBeacon","requestId":"1220578e-0b3d-4721-aa53-bd063055a111","gossip_listener_channel":"TBFollowingVotingGossip","sessionId":"ad8da44c-bfc2-4f6c-b13c-181193a570d2","errmsg":"verify following voting message: get node ATXID for epoch (miner ID d402d): atx for node d402d with publication epoch 33: leveldb: not found","name":"tBeacon"}
1495:{"L":"ERROR","T":"2021-09-14T07:09:40.001+0300","N":"d402d.blockOracle  ","M":"failed to calculate eligibility proofs","node_id":"d402d6a59d870552a7d9c7ca6af2d1fada8e189fea82a712aa5495b6019388c0","module":"blockOracle","layer_id":205,"epoch_id":34,"errmsg":"failed to get latest atx for node in epoch 34: failed to get atx id for target epoch 34: atx for node d402d with publication epoch 33: leveldb: not found","name":"blockOracle"}
```

## Steps to reproduce
Sync new node and see errors.

## Actual Behavior
Messages are sent without verifying eligibility.

## Expected Behavior
Eligibility should be verified before sending any messages.

### LINKED ISSUE #2837: tortoise beacon: different beacon calculated for the same epoch
## Description

running dockertest-late-nodes system test, one can see different beacon values were calculated.

Screenshot from 2021-09-28 13-58-19

the issue is that there is a race condition between the tortoise beacon gossip listener and the main protocol driver.
if the node is not participating in the beacon protocol (not synced), it still receives gossip messages for the protocol and pollutes the data structure for the next epoch.

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

Closes #2781
Closes #2837

## Changes

- check self ATX in previous epoch before participating in tortoise beacon and building blocks
- discard gossip messages when we are not in protocol

## Test Plan

unit tests

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: tortoisebeacon/tortoise_beacon.go [hunk 1]
```diff
@@ -348,6 +367,12 @@ func (tb *TortoiseBeacon) handleEpoch(ctx context.Context, epoch types.EpochID)
 		return
 	}
 
+	// make sure this node has ATX in the last epoch and is eligible to participate in tortoise beacon
+	if _, err := tb.atxDB.GetNodeAtxIDForEpoch(tb.nodeID, epoch-1); err != nil {
+		logger.Info("node has no ATX in last epoch. not participating in tortoise beacon")
```
Comment (nkryuchkov) [inline tortoisebeacon/tortoise_beacon.go]:
```suggestion
		logger.Info("node has no ATX in last epoch, not participating in tortoise beacon")
```

**Review by nkryuchkov — APPROVED** [review]

**Comment by countvonzero** [comment]
bors try

File: tortoisebeacon/tortoise_beacon.go (hunk 1 above)
Comment (countvonzero) [reply tortoisebeacon/tortoise_beacon.go]: done

**Comment by countvonzero** [comment]
Screenshot from 2021-09-28 18-43-50

in late-node system test with this PR, up to epoch 8, the last one in the system test, all nodes have the same beacon value

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**